core/src/main/java/com/linecorp/armeria/client/HttpSessionHandler.java

```diff
@@ -562,4 +579,44 @@ public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause) throws E
             // Netty will close the connection automatically on an IOException.
         }
     }
+
+    private void applyOutlierDetection(ClientRequestContext ctx) {
+        if (outlierDetectingRule == null || !isAcquirable()) {
+            return;
+        }
+        assert outlierDetector != null;
+        if (outlierDetector.isOutlier()) {
+            return;
+        }
+
+        ctx.log().whenComplete().thenAccept(log -> {
+            final OutlierDetectionDecision decision =
+                    outlierDetectingRule.decide(ctx, log.responseHeaders(), log.responseCause());
+            if (decision == null) {
+                return;
+            }
+            switch (decision) {
+                // NEXT is assumed as SUCCESS.
+                case SUCCESS:
+                case NEXT:
+                    outlierDetector.onSuccess();
+                    // The connection was marked as an outlier, but it's back to normal.
+                    if (outlierDetector.isOutlier()) {
+                        markUnacquirable();
```

Comment:
Note: I would still recommend users (especially those using HTTP1) to use a `RetryingClient` as I think it's possible that requests that unluckily acquired this connection could be failed due to this call depending on the timing. Although I know the code path to `markUnacquirable` is already exposed to the user, I feel like the timing of this call is more unpredictable with this change.

e.g. If `canSendRequest = false`:
https://github.com/line/armeria/blob/2219de19d510da43b7fc4afcabf730f8a4cca2ce/core/src/main/java/com/linecorp/armeria/client/AbstractHttpRequestHandler.java#L160

Reply:
Good point. What do you think of explicitly executing this callback in the EventLoop to guarantee the execution order?